### ISSUE TITLE
Umbrella issue for developer's guide updates

### ISSUE THREAD

**Issue by ryanj** [kubernetes/community#1919]
As per kubernetes/website#6102, and #1413 we need a developers' guide that is easily found, complete, and organized logically.

I'll plan on documenting next-steps in this umbrella issue.

**Comment by idvoretskyi** [comment]
cc @kubernetes/sig-docs-maintainers

**Comment by zacharysarah** [comment]
@idvoretskyi :+1:

**Comment by parispittman** [comment]
Thanks for this, @ryanj 

/sig contributor-experience

**Comment by ryanj** [comment]
Looking into these first:

- [ ] Establish a top-level Readme
- [ ] adding-an-APIGroup.md - Should be part of developer guide
- [ ] api-conventions.md - Should be part of developer guide
- [ ] api_changes.md - Should be part of developer guide
- [ ] architectural-roadmap.md - Should be part of developer guide and/or SIG Arch
- [ ] include links to examples and step-by-step guides

**Comment by heckj** [comment]
I think there would be a lot of value (and positive uptake, as well as inevitable bike-shedding) on having an example walk-through on  "how to set up a dev environment, compile the code, run the tests, and see the tests pass". This doesn't exist to my experience, or is heavily tribal knowledge, and may even vary quite significantly between subprojects within Kubernetes (i.e. kubernetes/helm vs. kubernetes/kubernetes vs. kubernetes/minikube).

**Comment by parispittman** [comment]
@carlpett - appreciate if you could list some of those missing pieces here so we can make sure to include/address. thanks for the help!

**Comment by carlpett** [comment]
Of course! 

This is in "chronological order of surprise" from the last few days :)

- *Import paths:* `k8s.io/X` maps to `github.com/kubernetes/kubernetes/staging/X`. This is then found by the compiler because there are symlinks in vendor/k8s.io. Unexpected to someone used to Golang. (This is documented in `staging/README.md`, but you'll only find that if you know what you are looking for)
- *Making changes:* There are a lot of places that need to be changed to make a change in a type, which seems to be required for many patches. After a while I got directed on Slack to `api_changes.md`, which covers it, but also _a lot_ of other things. Perhaps a split it out, or have a simple version near the top of the document? `api_changes.md` also isn't linked from any of the other documentation pages, so it is not easy to find (there are 57 .md files in that directory).
- *Generating files:* There are a lot of different file generation things going on. `make generated_files` does some of them, but there are also a number of other generator scripts under `./hack/`. This is also mentioned in `api_changes.md`, if you find it and read far enough.
- *Building:* There are no mentions of how to build in the devel/development.md file, except a note that you can experimentally build with Bazel. You need to find `build/README.md`, where `build/run.sh make` is introduced. (It is not clear what happens if you just run `make`. It seemed to hang on my machine, not sure if it is meant to work or not)
- *Running:* If you do not want to run it directly on your machine, directions are unclear. It is mentioned that it is possible, but not how. If you do want to run it directly on your machine, there's a pretty long list of requirements, but still lacks version information (etcd is commonly only available in 2.x versions in the default package managers, for example, but 3.x is required to start, so you'll need to install it somehow)
- *More building:* There are several other make targets that sometimes(?) need to be run. It's a bit confusing in in the docs when/if there should be run with `make TARGET` vs `./build/run.sh make TARGET`. Would be great with some clear guidelines here.

As a wishlist for how the docs could have been structured in a way that I would have found intuitive, something like this (sort of a step-by-step style):

- Prerequisites: Getting the code and development tools (not necessarily covering the "running" part here, just what you'll need to compile it)
- Navigating the code base (what components live where, the gotcha regarding k8s.io/vendoring, ...)
- Making a small change (some simple demonstration about what is required to do some typical first change)
- Building your change (could have this before the change part, but since just compiling your change will take ~10-15 minutes, there's a high risk of losing people if they have to do it twice..)
- Running your changed Kubernetes (might be merged with the previous one? Will get long, though...)
- Adding a test
- Making a PR
- (Quick start: Alternative route for the impatient/short summary for reference - commands required to get up and running, etc)

Having a few different tutorials/model PRs for making changes in different parts of the system might be a nice addition as well.

**Comment by ryanj** [comment]
Sorry for the delay on this issue!  I'm mostly blocked for another 2 weeks, but am hoping to loop back around to it afterwords. If anyone else wants to jump on this issue, feed free. I don't want to get in the way.

**Comment by parispittman** [comment]
/remove-lifecycle stale

we need to prioritize this and find an owner

**Comment by nikhita** [comment]
xref: https://github.com/kubernetes/community/issues/892 - this issue also has a lot of useful discussion regarding a developer guide.

**Comment by nikhita** [comment]
Related: @eduartua will be working on this as a part of [Outreachy](https://www.outreachy.org/) and has created an umbrella issue https://github.com/kubernetes/community/issues/3064 about this.

**Comment by nikhita** [comment]
I think we can close this issue in favour of https://github.com/kubernetes/community/issues/3064, since that is being actively worked on.

I'll leave this issue open until the end of the week and close it in case no one has any objections.

**Comment by nikhita** [comment]
Closing in favour of #3064. Please reopen if needed.

/close
